src/pybind/mgr/cephadm/module.py

```diff
@@ -2535,10 +2535,42 @@ def _apply(self, spec: GenericSpec) -> str:
 
         return self._apply_service_spec(cast(ServiceSpec, spec))
 
+    def _get_candidate_hosts(self, placement: PlacementSpec) -> List[str]:
+        """Return a list of candidate hosts according to the placement specification."""
+        all_hosts = self.cache.get_schedulable_hosts()
+        draining_hosts = [dh.hostname for dh in self.cache.get_draining_hosts()]
+        ls = []
+        if placement.hosts:
+            ls = [h.hostname for h in placement.hosts
+                  if h.hostname in placement.hosts and h.hostname not in draining_hosts]
+        elif placement.label:
+            ls = [x.hostname for x in [h for h in all_hosts if placement.label in h.labels]]
+        elif placement.host_pattern:
+            ls = [x for x in placement.filter_matching_hostspecs(all_hosts)]
+        elif (placement.count is not None or placement.count_per_host is not None):
+            ls = [x.hostname for x in all_hosts]
+        return ls
+
+    def _validate_placement_spec(self, spec: PlacementSpec) -> None:
+        if spec.count is not None:
+            raise OrchestratorError("Placement 'count' field is no supported for this specification.")
+        if spec.count_per_host is not None:
+            raise OrchestratorError("Placement 'count_per_host' field is no supported for this specification.")
+        if spec.hosts:
+            all_hosts = [h.hostname for h in self.inventory.all_specs()]
+            invalid_hosts = [h.hostname for h in spec.hosts if h.hostname not in all_hosts]
+            if invalid_hosts:
+                raise OrchestratorError(f"Found invalid host(s) in placement section: {invalid_hosts}. "
+                                        f"Please check 'ceph orch host ls' for available hosts.")
```

Comment:
we should probably note in this error message that we are excluding draining hosts as potential candidates to avoid confusion